src/inet/DNSResolver.cpp

```diff
@@ -298,12 +292,11 @@ INET_ERROR DNSResolver::Cancel()
  */
 void DNSResolver::HandleResolveComplete()
 {
+    Weave::System::Object::AutoRelease objectRelease(*this);
```

Comment:
I see this as an anti-pattern.  In this case we're dealing with a reference count, vs. a lock, where the lifetime of reference spans multiple states of the object.  In this case, a single call to a method called Release() is clearly the most obvious pattern.